core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java

```diff
@@ -127,8 +131,17 @@ protected void validate(TableMetadata base) {
       }
     }
 
+    // validating concurrently added data files is optional and is only needed to achieve serializable isolation
     if (conflictDetectionFilter != null && base.currentSnapshot() != null) {
       validateAddedDataFiles(base, startingSnapshotId, conflictDetectionFilter, caseSensitive);
     }
+
+    // validating concurrently added delete files is required whenever we overwrite specific data files
+    // if we find a new delete file matching one of the data files we are trying to overwrite,
+    // we must fail this operation as it would undelete rows that were removed concurrently
```

Comment:
Nit: it _may_ undelete rows that were removed concurrently. We don't really know because the overwrite might not.

(Actually, we could check this. If there are no added data files then it was a pure delete)

Reply:
I think we should automatically convert overwrites without added files to deletes.

Reply:
Fixed.